File: javascript/eclat.js
const { dataset7: transactions } = require('./core/data')

class Eclat {
  constructor ({ minsup = 2 }) {
    this.minsup = minsup
  }
  run (transactions) {
    const vert = this.mapToVerticalFormat(transactions)
    return this.mineFrequentItemsets(vert, 2)
  }
  uniqueItems (items) {
    return [...new Set(items)]
  }
  mapToVerticalFormat (transactions) {
    const vert = transactions.reduce((table, items, tid) =>
      this.uniqueItems(items).reduce((_table, item) => {
        _table[item] !== undefined ? _table[item].push(tid + 1) : _table[item] = [tid + 1]
        return _table
      }, table)
    , {})

    return Object.entries(vert).map(([item, tids]) => {
      return [[item], tids]
    })
  }
  union (arr1, arr2) {
    return this.uniqueItems([...arr1, ...arr2]).sort()
  }
  mineFrequentItemsets (transactions, k = 2) {
    const out = this.frequentItemsets(transactions, k)
    if (!out.length) return []

    return out.concat(this.mineFrequentItemsets(out, k + 1))
  }
  frequentItemsets (transactions, k = 2) {
    let output = []
    let found = {}
    transactions.forEach(([itemset1, tid1], i) => {
      transactions.forEach(([itemset2, tid2], j) => {
        if (i <= j) return
        const combinations = this.union(itemset1, itemset2)
        const tids = this.intersection(tid1, tid2)
        const key = combinations.join(',')
        if (combinations.length >= k && tids.length >= this.minsup && !found[key]) {
          found[key] = true
          output.push([combinations, tids])
        }
      })
    })
    return output
  }
  intersection (arr1, arr2) {
    const [target, compare] = arr1.length > arr2.length ? [arr1, new Set(arr2)] : [arr2, new Set(arr1)]
    return target.filter((item) => {
      return compare.has(item)
    })
  }
}

function main () {
  const eclat = new Eclat({ minsup: 2 })
  const model = eclat.run(transactions)
  console.log(model)
  // [ [ [ 'bread', 'butter' ], [ 1, 4, 8, 9 ] ],
  // [ [ 'bread', 'jam' ], [ 1, 8 ] ],
  // [ [ 'butter', 'jam' ], [ 1, 8 ] ],
  // [ [ 'butter', 'coke' ], [ 2, 4 ] ],
  // [ [ 'bread', 'milk' ], [ 5, 7, 8, 9 ] ],
  // [ [ 'butter', 'milk' ], [ 3, 6, 8, 9 ] ],
  // [ [ 'bread', 'butter', 'jam' ], [ 1, 8 ] ],
  // [ [ 'bread', 'butter', 'milk' ], [ 8, 9 ] ] ]
}

main()

File: haskell/eclat.hs
import qualified Data.Map as Map
import qualified Data.List as List
import qualified Data.Maybe as Maybe
import qualified Data.Set as Set

main :: IO()
main = do
  let minimumSupport = 2 -- minimumSupport is the minimum frequency the items should appear in each transactions
  let k = 2 -- k is the frequent itemsets count. if k is 2, them it should contain a minimum of two items e.g. {bread, butter}

  let transactions = [["bread", "butter", "jam"],
                      ["butter", "coke"],
                      ["butter", "milk"],
                      ["bread", "butter", "coke"],
                      ["bread", "milk"],
                      ["butter", "milk"],
                      ["bread", "milk"],
                      ["bread", "butter", "milk", "jam"],
                      ["bread", "butter", "milk"]]

  let verticalFormat = toVerticalFormat transactions -- convert the transactions from horizontal to vertical format
  print $ eclat verticalFormat minimumSupport k -- perform a recursive eclat to find all n-frequent itemsets

type Item = String
type ItemRow = [Item]

type ItemSet = Set.Set String
type TIDs = Set.Set Int
type VerticalFormatTransactions = Map.Map ItemSet TIDs

toVerticalFormat :: [ItemRow] -> VerticalFormatTransactions
toVerticalFormat transactions = 
  groupByItem $ concat [[(Set.fromList [item], Set.fromList [tid]) | item <- row ] | (row, tid) <- indexedItem]
  where 
    indexedItem = zip transactions [1..]
    groupByItem a =  Map.fromListWith (Set.union) a


frequentItemsets :: VerticalFormatTransactions -> Int -> Int -> VerticalFormatTransactions
frequentItemsets transactions minimumSupport k =
  aboveMinimumSupportThreshold . mapToList $ [(unionItems, intersectionItems) | 
                                              (item1, ids1) <- Map.toList transactions, 
                                              (item2, ids2) <- Map.toList transactions, 
                                              -- let minimumCombinations = minimum [Set.size item1, Set.size item2],
                                              let unionItems = item1 `Set.union` item2, 
                                              let intersectionItems = ids1 `Set.intersection` ids2, 
                                              Set.size unionItems >= k]
  where 
    aboveMinimumSupportThreshold = Map.filter (\items -> Set.size items >= minimumSupport)
    mapToList = Map.fromList

eclat :: VerticalFormatTransactions -> Int -> Int -> VerticalFormatTransactions
eclat transactions minimumSupport k
  | Map.null transactions = Map.empty 
  | otherwise = let output = frequentItemsets transactions minimumSupport k
                in Map.union output (eclat output minimumSupport (k + 1))
